flatpak-proxy.c

```diff
@@ -826,76 +833,124 @@ side_closed (ProxySide *side)
     }
 }
 
+/* This function is to avoid situations like this
+ *
+ * BEGIN\r\nl\0\0\1...
+ *
+ * e.g. where we read into the first D-Bus message while waiting for
+ * the final line from the client
+ */
+static gssize
+_my_g_socket_receive_message_safe (GSocket* socket,
+                                   GSocketAddress** address,
+                                   GInputVector* vector,
+                                   GSocketControlMessage*** messages,
+                                   gint* num_messages,
+                                   gint* flags,
+                                   GCancellable* cancellable,
+                                   GError** error)
+{
+  GString *str;
+  gchar c;
+  gssize num_read = 0;
+  gboolean last_was_cr;
+  GInputVector v;
+
+  str = g_string_new (NULL);
+
+  last_was_cr = FALSE;
+  v.buffer = &c;
+  v.size = 1;
+
+  /* receive one byte at a time until we have enough data for the vector 
+   * or we get \r\n
+   */
+  while (str->len < vector->size)
+    {
+      num_read = g_socket_receive_message (socket, address, &v, 1,
+                                           messages,
+                                           num_messages,
+                                           flags, cancellable, error);
+
+      if (num_read <= 0)
```

Comment:
This feels a bit wrong. We're reading N bytes, but then get an error, so we throw away the N bytes and report just the error. Typically network code doesn't want to lose data like that. However, maybe in this case it is fine, as on error we stop everything. Should have a comment about it though.

Reply:
Actually, this isn't quite true. The code does handle G_IO_ERROR_WOULD_BLOCK. So, you need to handle a partial read that returns a WOULD_BLOCK and return the partial results. And then you need to remember for the next call whether you have seen the CR.

Reply:
Honestly, Its probably best to just always return error only on the first byte read, and otherwise return what you read up to now.

Reply:
Good catch, I'll rework this.